### ISSUE TITLE
[Feature] Challenge healthcheck

### ISSUE THREAD

**Issue by reteps** [CTFd/CTFd#2594]
# Idea

On each challenge in CTFd, an indicator of challenge health. This idea would let you use CTFd as a 'budget' healthcheck system.

## Implementation

- A `challenges_healthcheck` function that expires every ~5 minutes (using a timed lru cache)
- An optional healthcheck field file reference on each challenge for a python file + requirements.txt
- Add a "healthcheck" docker container to the docker-compose.yml listening for requests
- When the `challenge_healthcheck` function expires, send a request to the internal "healthcheck" docker container containing all challenge files, connection info, etc.
- Create a venv for the request based on the hash of all inputs. E.g. `/runner/<hash>/{healthcheck.py, requirements.txt`
- Run the healthcheck script, passing in input files/connection info for the script to use. Use a chroot for each healthcheck?
- Return the exit code of the healthcheck back to the frontend

## Limitations

- All healthchecks run inside the same container
- All healthcheck scripts must be written in python

**Comment by reteps** [comment]
Curious if you have any thoughts @ColdHeat

**Comment by ColdHeat** [comment]
I think ultimately the healthcheck should be handled by a seperate service and CTFd can poll that. In your implementation how would you check the health of a plain TCP server?

**Comment by reteps** [comment]
So you would have your challenge associated with a plain TCP server.

Then the healthcheck is

```python
@route('/')
def check():
    return {status: <ping TCP server is successful>}
```

**Comment by pandatix** [comment]
With this proposal, how can you provide a healthcheck for a challenge that would require multiple services ?
Like a small-business infrastructure instanciated on demand, how to monitor each service is available and not only one ?

+1 on the Separation of Concerns Principle: to me it should be the goal of another Constituent System of a CTF infrastructure to handle infrastructures, CTFd should only be the gate to call those services through a plugin.

**Comment by reteps** [comment]
> how can you provide a healthcheck for a challenge that would require multiple services ?

This is a good point, a single string isn't enough to healthcheck e.g. a challenge with an admin bot and a challenge page. However I have never seen a challenge with more than 2 remotes needed.

> how to monitor each service is available and not only one ?

The idea of this healthcheck is more of a **solvecheck** -- so I am not interested in talking to the challenges internal DB, etc. but rather if the challenge is solvable or not from the information in the challenge description.

> CTFd should only be the gate to call those services through a plugin.

I like the plugin idea if this isn't something CTFd wants to support upstream. @ColdHeat let me know your opinion, as I will either be building out a PR to CTFd or a plugin for this.

**Comment by pandatix** [comment]
> The idea of this healthcheck is more of a solvecheck

Ok, this may require additional clarification then (e.g., in a doc), it could be interpreted differently depending on the reader's background :smile:
Nevertheless, I don't know how far you could go with a solvecheck : what if it affects the state of the challenge ? I think you consider the challenge to be stateless, but if solving a challenge requires shooting down a protection mecanism, then the player can work on a completely different challenge.

Moreover, it also considers that the whole `Challenge` object is related to infrastructure, which is only a flavor of challenge : it could be a quiz, a file(s)-based one, an infra-based one, etc.

**Comment by reteps** [comment]
>  if solving a challenge requires shooting down a protection mecanism, then the player can work on a completely different challenge

For a challenge solvable by multiple people, having a non-stateless challenge doesn't seem like a good idea.


> a quiz

Shouldn't need a healthcheck?

> a file(s)-based one

You can write a healthcheck that uses the deployed files and not any connection info

> infra-based one

I don't see why this wouldn't adapt, e.g. to a cloud deployed challenge

**Comment by ColdHeat** [comment]
I’m not against this but it would be faster for you to implement it as a
plug-in and then we can see how it works in practice before changing things
here in the upstream.

I do feel like uptime checking is better left to another service instead of
making it CTFd’s responsibility.

**Comment by reteps** [comment]
Sounds good, I'll update this issue/the PR with more details when it is deployed as a plugin.

**Comment by reteps** [comment]
> I do feel like uptime checking is better left to another service instead of
making it CTFd’s responsibility.

Just to be clear, the idea here is that you have another healthcheck server that is polling the actual uptime, and CTFd is just reading that state. Minor distinction, but I think it's important.

**Comment by pandatix** [comment]
@reteps sorry if I may have sound unclear... :sweat_smile: 
I was speaking of a root object `Challenge` specialized for infrastructures. A basic `Challenge` object, as part of CTFd API, is handling generic attributes, and plugins can then extend them through a plugin, with or without mutations. Those changes provide the functionalities required by your use case to fulfill your goals.
This specialization seems more of a plugin to me rather than CTFd in itself, has CTFd is not made for deploying infrastructures on demand thus is not made for monitoring things out of itself : this is a good separation of concerns principle example :)
An example of specialization follows.

```mermaid
classDiagram
    class BaseChallenge{
        +int id
        +string name
        +string description
        +string category
        +int max_attempts
        +int value
        ...
    }

    DynamicChallenge --|> BaseChallenge: extends
    class DynamicChallenge{
        +int initial
        +int decay
        +int minimum
        +string function
    }

    HealthcheckChallenge --|> DynamicChallenge: extends
    class HealthcheckChallenge{
        +string solve_path
    }

    QuizzChallenge --|> BaseChallenge: extends
    class QuizzChallenge{
        +int combo
    }
    QuizzChallenge "*" *--> QuizzQuestion: questions
    class QuizzQuestion{
        +string question
        +string answer
    }
```

> Just to be clear, the idea here is that you have another healthcheck server that is polling the actual uptime, and CTFd is just reading that state. Minor distinction, but I think it's important.

Ok then this sounds like a Service Oriented Architecture: CTFd displays the info to end-users (players, admins), the heathcheck service monitors infrastructure parts (e.g., for solvability). That may fit a microservices architecture.

**Comment by reteps** [comment]
> That may fit a microservices architecture.

Exactly. I assume the CTF is run with a microservice architecture, with each challenge deployed independently.
